src/core/fetcher.ts

```diff
@@ -14,12 +14,32 @@ import { CompletionItem, CompletionItemKind, CompletionList, workspace, window }
 import { sortText } from "../providers/autoCompletion";
 import { CrateMetadatas } from "../api/crateMetadatas";
 
-export async function fetchCrateVersions(dependencies: Item[]): Promise<[Promise<Dependency[]>, Map<string, Dependency[]>]> {
+export async function fetchCrateVersions(dependencies: Item[]): Promise<[Dependency[], Map<string, Dependency[]>]> {
   // load config
   const config = workspace.getConfiguration("");
   const shouldListPreRels = !!config.get("crates.listPreReleases");
-  var indexServerURL = config.get<string>("crates.indexServerURL") ?? sparseIndexServerURL;
+  const indexServerURL = config.get<string>("crates.indexServerURL") ?? sparseIndexServerURL;
+  const otherIndexServerURLs = config.get<string[]>("crates.otherIndexServerURLs") ?? sparseIndexServerURL;
+  const allServerUrls = [indexServerURL, ...otherIndexServerURLs];
 
+  let responsesMap: Map<string, Dependency[]> = new Map();
+  const responses = await Promise.all(allServerUrls.map(url => fetchCrateVersionsFromIndex(dependencies, shouldListPreRels, url)));
+
+  return [responses.reduce((accumulatedDependencies, current) => {
+    if (accumulatedDependencies.length === 0) {
+      return current;
+    }
+    for (let i = 0; i !== accumulatedDependencies.length; i++) {
+      if (accumulatedDependencies[i].error) {
+        // this dependency failed so far, replaced by current
+        accumulatedDependencies[i] = current[i];
+      }
+    }
+    return accumulatedDependencies;
+  }, []), responsesMap];
```

Comment:
This doesn't really specify it, but like the logic here is, if this is the first dependency, add it to the accumulator. For all remaining items in the input, skip them, but if the only item we have _is_ an error, then replace it with that one.

If this is the intended logic, then the entire reduce may be more simply handled by a single for loop over `responses`, selecting the first element that is not an error.

```suggestion
let dependencies = [];
for (let i = 0; i < responses.length; i++) {
  if (!responses[i].error) {
    dependency.push(responses[i]);
    break;
  }
}

// If all responses were errors, at least include the first response so the error gets reported.
if (dependencies.length === 0) {
  dependencies.push(responses[0]);
}

return [dependencies, responseMap];
```

If the intended behavior is different, because we want to add multiple successful dependencies, then you current iteration is still a bit more complicated than necessary, and could be achieved by removing the `break` above.

Reply:
Thanks for the feedback! I think the `reduce` was not really clear as to what it does. The logic is that, at this point we have all the responses from all the servers in the `responses` variables. The type of `responses` is `Dependency[][]`. In the new proposed version, we first take the dependencies from the first server, that is usually crates.io. And then for the responses for other servers we accumulate by replacing the dependencies that are still errors.